core/sql/optimizer/BindRelExpr.cpp

```diff
@@ -6807,6 +6846,100 @@ RelExpr *RelRoot::bindNode(BindWA *bindWA)
+  // where rn <= n
+  //
+  // Aside: For row subqueries in the top-most select list, the 
+  // chicken-and-egg problem doesn't exist because the top-most RelRoot
+  // does survive until normalizeNode time.
+  //
+  // To perform this transformation, we take the input tree (on the left
+  // below) and rewrite it to the output tree (on the right below). Since
+  // this transformation is being done before binding, we move around
+  // ItemExpr trees instead of ValueIds. Binding happens at the end of
+  // the method, after the rewrite.
+  //
+  //   
+  //                                       new RelRoot (newRelRoot), with copy of
+  //                                        original select list
+  //                                          | 
+  //                                       new RenameTable (renameTable), 
+  //                                         with the WHERE rn <= n
+  //                                         clause attached
+  //                                          |
+  //     RelRoot of subquery (this)        original RelRoot (this),
+  //       with original select             with ROW_NUMBER ORDER BY aggregate added
+  //       list (originalSelectList)         to select list, ORDER BY removed 
+  //        |                                from RelRoot 
+  //        |                                 |
+  //        |                              new RelSequence (sequenceNode)
+  //        |                                 |
+  //     Subquery tree (query)             Subquery tree (query)            
+
+
+  // point to subquery tree
+  RelExpr * query = child(0)->castToRelExpr();
+
+  // retain a pointer to the present select list to put in our new RelRoot
+  ItemExpr * originalSelectList = getCompExprTree();
+
+  // create "row_number() over(order by <y>) as rn" parse tree, removing the
+  // order by tree from this RelRoot in the process
+  Aggregate * rowNumberOverOrderBy = 
+    new (bindWA->wHeap()) Aggregate(ITM_COUNT, 
+                                    new (bindWA->wHeap()) SystemLiteral(1),
+                                    FALSE /*i.e. not distinct*/,
+                                    ITM_COUNT_STAR__ORIGINALLY,
+                                    '!');
+  rowNumberOverOrderBy->setOLAPInfo(NULL, removeOrderByTree(), -INT_MAX, 0);
+  NAString rowNumberColumnName = "_sys_RN_" + bindWA->fabricateUniqueName();
+  ColRefName * colRefName = new (bindWA->wHeap()) ColRefName(rowNumberColumnName, bindWA->wHeap());
+  RenameCol * rename = new (bindWA->wHeap())RenameCol(rowNumberOverOrderBy, colRefName);
+
+  // add it to select list of the current RelRoot
+  compExprTree_ = new (bindWA->wHeap()) ItemList(compExprTree_,rename);
+
+  // put a RelSequence node on top of the query node, and make it the child of this
+  RelSequence * sequenceNode = new (bindWA->wHeap()) RelSequence(query,NULL);
+  sequenceNode->setHasOlapFunctions(TRUE);
+  this->child(0) = sequenceNode;
```

Comment:
Does the sequence node sends a GET_ALL or GET_N to it's child ?

Reply:
Also what happens if there are ESPs How does the plan look ?

Reply:
Looking at the code in executor/ExSequence.cpp, it looks like it sends GET_ALL. It is at the sequence node that the data is ordered and the OLAP function is calculated. We have to do a GET_ALL in order to get the ordering.

Reply:
Re: ESPs: I will research and find out. If I recall correctly, sequence plans were not easy to parallelize. But that recollection is many years old. It may have improved since.

Reply:
> Looking at the code in executor/ExSequence.cpp, it looks like it sends GET_ALL. It is at the sequence node that the data is ordered and the OLAP function is calculated. We have to do a GET_ALL in order to get the ordering.
Ok so if that's the case, then with this first N in subquery, we won't be getting any performance  , correct ? It will work correctly but wondering if the user will expect it to run  faster  than if they did not provide the first N ?

Reply:
Hmmmm... The presence of an ORDER BY implies that all rows must be read and sorted. There is no escaping that. Even in firstn plans (where [first n] + ORDER BY is specified at the top level query) we do that. There may be an optimization possibility of doing Top N sort rather than a full sort. But in any case, all rows must be read.

Reply:
Re: ESPs (again): I just did an example with two tables of 2 million rows each, and forced ESP parallelism. The sequence operator is done serially, in one ESP. So my recollections were correct, we do not parallelize OLAP operators today.

Reply:
I should add that in my experiment, all other parts of the plan were parallelized. So for example, the scans underneath the sequence operator were done in parallel, and each was sorted in parallel, with the results funneled into the (serialized) sequence operator.